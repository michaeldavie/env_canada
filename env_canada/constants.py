USER_AGENT = "env_canada/0.6.2"
