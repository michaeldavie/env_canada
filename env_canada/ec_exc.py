class UnknownStationId(Exception):
    """Station ID is unknown"""
