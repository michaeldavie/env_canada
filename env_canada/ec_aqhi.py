from datetime import datetime, timezone
import logging

from aiohttp import ClientSession
import defusedxml.ElementTree as et
from geopy import distance
import voluptuous as vol

from .constants import USER_AGENT

AQHI_SITE_LIST_URL = "https://dd.weather.gc.ca/air_quality/doc/AQHI_XML_File_List.xml"
AQHI_OBSERVATION_URL = "https://dd.weather.gc.ca/air_quality/aqhi/{}/observation/realtime/xml/AQ_OBS_{}_CURRENT.xml"
AQHI_FORECAST_URL = "https://dd.weather.gc.ca/air_quality/aqhi/{}/forecast/realtime/xml/AQ_FCST_{}_CURRENT.xml"

LOG = logging.getLogger(__name__)

ATTRIBUTION = {
    "EN": "Data provided by Environment Canada",
    "FR": "Donn√©es fournies par Environnement Canada",
}


__all__ = ["ECAirQuality"]


def timestamp_to_datetime(timestamp):
    dt = datetime.strptime(timestamp, "%Y%m%d%H%M%S")
    dt = dt.replace(tzinfo=timezone.utc)
    return dt


async def get_aqhi_regions(language):
    """Get list of all AQHI regions from Environment Canada, for auto-config."""
    zone_name_tag = "name_%s_CA" % language.lower()
    region_name_tag = "name%s" % language.title()

    LOG.debug("get_aqhi_regions() started")

    regions = []
    async with ClientSession(raise_for_status=True) as session:
        response = await session.get(
            AQHI_SITE_LIST_URL, headers={"User-Agent": USER_AGENT}, timeout=10
        )
        result = await response.read()

    site_xml = result.decode("utf-8")
    xml_object = et.fromstring(site_xml)

    for zone in xml_object.findall("./EC_administrativeZone"):
        _zone_attribs = zone.attrib
        _zone_attrib = {
            "abbreviation": _zone_attribs["abreviation"],
            "zone_name": _zone_attribs[zone_name_tag],
        }
        for region in zone.findall("./regionList/region"):
            _region_attribs = region.attrib

            _region_attrib = {
                "region_name": _region_attribs[region_name_tag],
                "cgndb": _region_attribs["cgndb"],
                "latitude": float(_region_attribs["latitude"]),
                "longitude": float(_region_attribs["longitude"]),
            }
            _children = list(region)
            for child in _children:
                _region_attrib[child.tag] = child.text
            _region_attrib.update(_zone_attrib)
            regions.append(_region_attrib)

    LOG.debug("get_aqhi_regions(): found %d regions", len(regions))

    return regions


async def find_closest_region(language, lat, lon):
    """Return the AQHI region and site ID of the closest site."""
    region_list = await get_aqhi_regions(language)

    def site_distance(site):
        """Calculate distance to a region."""
        return distance.distance((lat, lon), (site["latitude"], site["longitude"]))

    return min(region_list, key=site_distance)


class ECAirQuality(object):

    """Get air quality data from Environment Canada."""

    def __init__(self, **kwargs):
        """Initialize the data object."""

        init_schema = vol.Schema(
            vol.All(
                vol.Any(
                    {
                        vol.Required("coordinates"): object,
                        vol.Optional("language"): object,
                    },
                    {
                        vol.Required("zone_id"): object,
                        vol.Required("region_id"): object,
                        vol.Optional("language"): object,
                    },
                ),
                {
                    vol.Optional("zone_id"): vol.In(
                        ["atl", "ont", "pnr", "pyr", "que"]
                    ),
                    vol.Optional("region_id"): vol.All(str, vol.Length(5)),
                    vol.Optional("coordinates"): (
                        vol.All(vol.Or(int, float), vol.Range(-90, 90)),
                        vol.All(vol.Or(int, float), vol.Range(-180, 180)),
                    ),
                    vol.Optional("language", default="EN"): vol.In(["EN", "FR"]),
                },
            )
        )

        kwargs = init_schema(kwargs)

        self.language = kwargs["language"]

        if (
            "zone_id" in kwargs
            and "region_id" in kwargs
            and kwargs["zone_id"] is not None
            and kwargs["region_id"] is not None
        ):
            self.zone_id = kwargs["zone_id"]
            self.region_id = kwargs["region_id"].upper()
        else:
            self.zone_id = None
            self.region_id = None
            self.coordinates = kwargs["coordinates"]

        self.metadata = {"attribution": ATTRIBUTION[self.language]}
        self.region_name = None
        self.current = None
        self.current_timestamp = None
        self.forecasts = dict(daily={}, hourly={})

    async def get_aqhi_data(self, url):
        async with ClientSession(raise_for_status=True) as session:
            try:
                response = await session.get(
                    url.format(self.zone_id, self.region_id),
                    headers={"User-Agent": USER_AGENT},
                    timeout=10,
                )
            except Exception:
                LOG.debug("Retrieving AQHI failed", exc_info=True)
                return None

            result = await response.read()
            aqhi_xml = result.decode("ISO-8859-1")
            return et.fromstring(aqhi_xml)

    async def update(self):
        # Find closest site if not identified

        if not (self.zone_id and self.region_id):
            closest = await find_closest_region(self.language, *self.coordinates)
            self.zone_id = closest["abbreviation"]
            self.region_id = closest["cgndb"]
            LOG.debug(
                "update() closest region returned: zone_id '%s' region_id '%s'",
                self.zone_id,
                self.region_id,
            )

        # Fetch current measurement
        aqhi_current = await self.get_aqhi_data(url=AQHI_OBSERVATION_URL)

        if aqhi_current:
            # Update region name
            element = aqhi_current.find("region")
            self.region_name = element.attrib[
                "name{lang}".format(lang=self.language.title())
            ]
            self.metadata["location"] = self.region_name

            # Update AQHI current condition
            element = aqhi_current.find("airQualityHealthIndex")
            if element is not None:
                self.current = float(element.text)
            else:
                self.current = None

            element = aqhi_current.find("./dateStamp/UTCStamp")
            if element is not None:
                self.current_timestamp = timestamp_to_datetime(element.text)
            else:
                self.current_timestamp = None
            self.metadata["timestamp"] = self.current_timestamp
            LOG.debug(
                "update(): aqhi_current %d timestamp %s",
                self.current,
                self.current_timestamp,
            )

        # Update AQHI forecasts
        aqhi_forecast = await self.get_aqhi_data(url=AQHI_FORECAST_URL)

        if aqhi_forecast:
            # Update AQHI daily forecasts
            for f in aqhi_forecast.findall("./forecastGroup/forecast"):
                for p in f.findall("./period"):
                    if self.language == p.attrib["lang"]:
                        period = p.attrib["forecastName"]
                self.forecasts["daily"][period] = int(
                    f.findtext("./airQualityHealthIndex") or 0
                )

            # Update AQHI hourly forecasts
            for f in aqhi_forecast.findall("./hourlyForecastGroup/hourlyForecast"):
                self.forecasts["hourly"][
                    timestamp_to_datetime(f.attrib["UTCTime"])
                ] = int(f.text or 0)
