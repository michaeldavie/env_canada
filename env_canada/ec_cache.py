from aiohttp import ClientSession
from datetime import datetime, timedelta

from .constants import USER_AGENT

CACHE_EXPIRE_TIME = timedelta(minutes=200)  # Time is tuned for 3h radar image


class CacheClientSession(ClientSession):
    """Shim to cache ClientSession requests."""

    _cache = {}

    def _flush_cache(self):
        """Flush expired cache entries."""

        now = datetime.now()
        expired = [key for key, value in self._cache.items() if value[0] < now]
        for key in expired:
            del self._cache[key]

    async def get(self, url, params, cache_time=CACHE_EXPIRE_TIME):
        """Thin wrapper around ClientSession.get to cache responses."""

        self._flush_cache()  # Flush at start so we don't use expired entries

        cache_key = (url, tuple(sorted(params.items())))
        result = self._cache.get(cache_key)
        if not result:
            result = (
                datetime.now() + cache_time,
                await super().get(
                    url=url, params=params, headers={"User-Agent": USER_AGENT}
                ),
            )
            self._cache[cache_key] = result

        return result[1]
