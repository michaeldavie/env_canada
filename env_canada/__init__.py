from .ec_aqhi import ECAirQuality
from .ec_historical import ECHistorical, ECHistoricalRange
from .ec_hydro import ECHydro
from .ec_radar import ECRadar
from .ec_weather import ECWeather
