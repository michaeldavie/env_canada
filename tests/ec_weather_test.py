import asyncio
from unittest.mock import AsyncMock, patch

import pytest
from freezegun import freeze_time
from syrupy.assertion import SnapshotAssertion

from env_canada import ECWeather, ec_weather


@pytest.mark.parametrize(
    "init_parameters",
    [
        {"coordinates": (50, -100)},
        {"station_id": "ON/s0000430"},
        {"station_id": "s0000430"},
        {"station_id": "430"},
    ],
)
def test_ecweather(init_parameters):
    weather = ECWeather(**init_parameters)
    assert isinstance(weather, ECWeather)


def setup_test(args) -> tuple[ECWeather, AsyncMock]:
    resp = AsyncMock(status_code=200)
    side_effects = []

    # For new URL structure, we need to mock both directory listing and weather file
    # First response: sites CSV
    # Second response: directory listing HTML
    # Third response: weather XML

    for k, v in args.items():
        if k == "sites":
            with open(v) as file:
                side_effects.append(file.read())
        elif k == "forecast":
            # Add mock directory listing HTML before the forecast
            station = args.get("station", "ON/s0000430")
            if "/" in station:
                _, station_num = station.split("/")
                station_num = station_num.replace("s0000", "")
            else:
                station_num = (
                    station.replace("s0000", "") if "s0000" in station else station
                )

            # Mock directory listing HTML with a sample file
            mock_html = f"""<html><body>
            <a href="20250206T120000.000Z_MSC_CitypageWeather_s0000{station_num.zfill(3)}_en.xml">file</a>
            </body></html>"""
            side_effects.append(mock_html)

            # Add the actual forecast XML
            with open(v) as file:
                side_effects.append(file.read())
        elif k == "exception":
            side_effects.append(v)
        elif k == "status_code":
            resp.status_code = v

    resp.text.side_effect = side_effects

    return (ECWeather(station_id=args["station"]), resp)


@pytest.mark.asyncio
async def test_weather_retrieved_weather_updates_ok(snapshot: SnapshotAssertion):
    ecw, resp = setup_test(
        {
            "station": "ON/s0000430",
            "sites": "tests/fixtures/site_list.csv",
            "forecast": "tests/fixtures/weather.xml",
        }
    )

    with patch("aiohttp.ClientSession.get", AsyncMock(return_value=resp)):
        with freeze_time("2025-02-06 00:00"):
            await ecw.update()

    assert ecw == snapshot


@pytest.mark.asyncio
async def test_weather_exception_on_old_forecast_data():
    ecw, resp = setup_test(
        {
            "station": "ON/s0000430",
            "sites": "tests/fixtures/site_list.csv",
            "forecast": "tests/fixtures/weather.xml",
        }
    )

    with patch("aiohttp.ClientSession.get", AsyncMock(return_value=resp)):
        with freeze_time("2025-02-06 03:00"):
            with pytest.raises(ec_weather.ECWeatherUpdateFailed):
                await ecw.update()


@pytest.mark.asyncio
async def test_weather_exception_returns_cached_data():
    ecw, resp = setup_test(
        {
            "station": "ON/s0000430",
            "sites": "tests/fixtures/site_list.csv",
            "forecast": "tests/fixtures/weather.xml",
        }
    )

    # Do the initial update with a good response
    with patch("aiohttp.ClientSession.get", AsyncMock(return_value=resp)):
        with freeze_time("2025-02-06 00:00"):
            await ecw.update()

    with patch("aiohttp.ClientSession.get", side_effect=TimeoutError):
        with freeze_time("2025-02-06 00:00"):
            await ecw.update()

    assert ecw.metadata.cache_returned_on_update == 1

    with patch("aiohttp.ClientSession.get", side_effect=TimeoutError):
        with freeze_time("2025-02-06 00:00"):
            await ecw.update()

    assert ecw.metadata.cache_returned_on_update == 2

    # Move date into future, should not return cached data now
    with patch("aiohttp.ClientSession.get", side_effect=TimeoutError):
        with freeze_time("2025-02-06 11:42"):
            with pytest.raises(ec_weather.ECWeatherUpdateFailed):
                await ecw.update()

    assert ecw.metadata.cache_returned_on_update == 0


@pytest.mark.slow
def test_get_ec_sites():
    sites = asyncio.run(ec_weather.get_ec_sites())
    assert len(sites) > 0


@pytest.mark.slow
def test_update_ec_weather():
    ecw, _ = setup_test({"station": "ON/s0000430"})
    asyncio.run(ecw.update())
    assert ecw.conditions


@pytest.mark.parametrize(
    "station_input,expected_station_number",
    [
        ("ON/s0000430", "430"),
        ("s0000430", "430"),
        ("430", "430"),
        ("1", "1"),
        ("99", "99"),
    ],
)
def test_validate_station(station_input, expected_station_number):
    """Test that station validation returns correct 3-digit format."""
    result = ec_weather.validate_station(station_input)
    assert result == expected_station_number


@pytest.mark.asyncio
async def test_station_id_formats_create_tuples():
    """Test that different station ID formats result in proper tuples."""
    test_cases = [
        ("ON/s0000430", ("ON", "430")),
        ("s0000430", ("ON", "430")),  # Should find ON from site data
        ("430", ("ON", "430")),  # Should find ON from site data
    ]

    for station_input, expected_tuple in test_cases:
        ecw, resp = setup_test(
            {
                "station": station_input,
                "sites": "tests/fixtures/site_list.csv",
                "forecast": "tests/fixtures/weather.xml",
            }
        )

        with patch("aiohttp.ClientSession.get", AsyncMock(return_value=resp)):
            with freeze_time("2025-02-06 00:00"):
                await ecw.update()

        assert ecw.station_id == expected_tuple
        assert ecw.lat is not None
        assert ecw.lon is not None
