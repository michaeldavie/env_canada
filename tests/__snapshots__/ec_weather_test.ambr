# serializer version: 1
# name: test_weather_retrieved_weather_updates_ok
  ECWeather(
    alerts=dict({
      'advisories': dict({
        'label': 'Advisories',
        'value': list([
          dict({
            'date': 'Wednesday February 05, 2025 at 18:07 EST',
            'title': 'Weather Advisory',
          }),
        ]),
      }),
      'endings': dict({
        'label': 'Endings',
        'value': list([
        ]),
      }),
      'statements': dict({
        'label': 'Statements',
        'value': list([
        ]),
      }),
      'warnings': dict({
        'label': 'Warnings',
        'value': list([
        ]),
      }),
      'watches': dict({
        'label': 'Watches',
        'value': list([
        ]),
      }),
    }),
    conditions=dict({
      'condition': dict({
        'label': 'Condition',
        'value': 'Mainly Clear',
      }),
      'dewpoint': dict({
        'label': 'Dew Point',
        'unit': 'C',
        'value': -21.1,
      }),
      'high_temp': dict({
        'label': 'High Temperature',
        'unit': 'C',
        'value': -4,
      }),
      'humidex': dict({
        'label': 'Humidex',
        'value': None,
      }),
      'humidity': dict({
        'label': 'Humidity',
        'unit': '%',
        'value': 53,
      }),
      'icon_code': dict({
        'label': 'Icon Code',
        'value': '31',
      }),
      'low_temp': dict({
        'label': 'Low Temperature',
        'unit': 'C',
        'value': -19,
      }),
      'normal_high': dict({
        'label': 'Normal High Temperature',
        'unit': 'C',
        'value': -5,
      }),
      'normal_low': dict({
        'label': 'Normal Low Temperature',
        'unit': 'C',
        'value': -14,
      }),
      'observationTime': dict({
        'label': 'Observation Time',
        'value': FakeDatetime(2025, 2, 5, 23, 0, tzinfo=tzutc()),
      }),
      'pop': dict({
        'label': 'Chance of Precip.',
        'value': None,
      }),
      'pressure': dict({
        'label': 'Pressure',
        'unit': 'kPa',
        'value': 103.0,
      }),
      'sunrise': dict({
        'label': 'Sunrise',
        'value': FakeDatetime(2025, 2, 5, 12, 19, tzinfo=tzutc()),
      }),
      'sunset': dict({
        'label': 'Sunset',
        'value': FakeDatetime(2025, 2, 5, 22, 15, tzinfo=tzutc()),
      }),
      'temperature': dict({
        'label': 'Temperature',
        'unit': 'C',
        'value': -13.6,
      }),
      'tendency': dict({
        'label': 'Tendency',
        'unit': 'kPa',
        'value': 'falling',
      }),
      'text_summary': dict({
        'label': 'Forecast',
        'value': 'Tonight. Clear. Becoming partly cloudy after midnight. Wind up to 15 km/h. Low minus 19. Wind chill near minus 26.',
      }),
      'uv_index': dict({
        'label': 'UV Index',
        'value': 1,
      }),
      'visibility': dict({
        'label': 'Visibility',
        'unit': 'km',
        'value': 24.1,
      }),
      'wind_bearing': dict({
        'label': 'Wind Bearing',
        'unit': 'degrees',
        'value': 142,
      }),
      'wind_chill': dict({
        'label': 'Wind Chill',
        'value': -18,
      }),
      'wind_dir': dict({
        'label': 'Wind Direction',
        'value': 'SE',
      }),
      'wind_gust': dict({
        'label': 'Wind Gust',
        'value': None,
      }),
      'wind_speed': dict({
        'label': 'Wind Speed',
        'unit': 'km/h',
        'value': 7,
      }),
    }),
    daily_forecasts=list([
      dict({
        'icon_code': '31',
        'period': 'Wednesday night',
        'precip_probability': 0,
        'temperature': -19,
        'temperature_class': 'low',
        'text_summary': 'Clear. Becoming partly cloudy after midnight. Wind up to 15 km/h. Low minus 19. Wind chill near minus 26.',
        'timestamp': FakeDatetime(2025, 2, 5, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '17',
        'period': 'Thursday',
        'precip_probability': 0,
        'temperature': -4,
        'temperature_class': 'high',
        'text_summary': 'Cloudy. Snow beginning in the morning. Risk of freezing drizzle late in the afternoon. Amount 5 to 10 cm. Wind up to 15 km/h. High minus 4. Wind chill minus 24 in the morning and minus 10 in the afternoon. UV index 1 or low.',
        'timestamp': FakeDatetime(2025, 2, 6, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '38',
        'period': 'Thursday night',
        'precip_probability': 60,
        'temperature': -7,
        'temperature_class': 'low',
        'text_summary': 'Snow ending early in the evening then mainly cloudy with 60 percent chance of flurries. Risk of freezing drizzle early in the evening. Risk of snow squalls overnight. Wind becoming west 30 km/h gusting to 50 in the evening. Low minus 7. Wind chill minus 8 in the evening and minus 16 overnight.',
        'timestamp': FakeDatetime(2025, 2, 6, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '08',
        'period': 'Friday',
        'precip_probability': 40,
        'temperature': -3,
        'temperature_class': 'high',
        'text_summary': 'A mix of sun and cloud with 40 percent chance of flurries. Windy. High minus 3.',
        'timestamp': FakeDatetime(2025, 2, 7, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '38',
        'period': 'Friday night',
        'precip_probability': 30,
        'temperature': -12,
        'temperature_class': 'low',
        'text_summary': 'Cloudy periods with 30 percent chance of flurries. Low minus 12.',
        'timestamp': FakeDatetime(2025, 2, 7, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '00',
        'period': 'Saturday',
        'precip_probability': 0,
        'temperature': -5,
        'temperature_class': 'high',
        'text_summary': 'Sunny. High minus 5.',
        'timestamp': FakeDatetime(2025, 2, 8, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '17',
        'period': 'Saturday night',
        'precip_probability': 0,
        'temperature': -10,
        'temperature_class': 'low',
        'text_summary': 'Snow. Low minus 10.',
        'timestamp': FakeDatetime(2025, 2, 8, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '16',
        'period': 'Sunday',
        'precip_probability': 0,
        'temperature': -4,
        'temperature_class': 'high',
        'text_summary': 'Periods of snow. High minus 4.',
        'timestamp': FakeDatetime(2025, 2, 9, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '32',
        'period': 'Sunday night',
        'precip_probability': 0,
        'temperature': -18,
        'temperature_class': 'low',
        'text_summary': 'Cloudy periods. Low minus 18.',
        'timestamp': FakeDatetime(2025, 2, 9, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '02',
        'period': 'Monday',
        'precip_probability': 0,
        'temperature': -8,
        'temperature_class': 'high',
        'text_summary': 'A mix of sun and cloud. High minus 8.',
        'timestamp': FakeDatetime(2025, 2, 10, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '32',
        'period': 'Monday night',
        'precip_probability': 0,
        'temperature': -15,
        'temperature_class': 'low',
        'text_summary': 'Cloudy periods. Low minus 15.',
        'timestamp': FakeDatetime(2025, 2, 10, 21, 40, tzinfo=tzutc()),
      }),
      dict({
        'icon_code': '02',
        'period': 'Tuesday',
        'precip_probability': 0,
        'temperature': -7,
        'temperature_class': 'high',
        'text_summary': 'A mix of sun and cloud. High minus 7.',
        'timestamp': FakeDatetime(2025, 2, 11, 21, 40, tzinfo=tzutc()),
      }),
    ]),
    forecast_time=FakeDatetime(2025, 2, 5, 21, 40, tzinfo=tzutc()),
    hourly_forecasts=list([
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 0, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -16,
        'wind_direction': 'S',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 1, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -17,
        'wind_direction': 'S',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 2, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -18,
        'wind_direction': 'S',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 3, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -19,
        'wind_direction': 'S',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 4, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -19,
        'wind_direction': 'S',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 5, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -19,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Clear',
        'icon_code': '30',
        'period': FakeDatetime(2025, 2, 6, 6, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -19,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'A few clouds',
        'icon_code': '31',
        'period': FakeDatetime(2025, 2, 6, 7, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -19,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'A few clouds',
        'icon_code': '31',
        'period': FakeDatetime(2025, 2, 6, 8, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -18,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'A few clouds',
        'icon_code': '31',
        'period': FakeDatetime(2025, 2, 6, 9, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -18,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Partly cloudy',
        'icon_code': '32',
        'period': FakeDatetime(2025, 2, 6, 10, 0, tzinfo=tzutc()),
        'precip_probability': 0,
        'temperature': -18,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Mainly cloudy',
        'icon_code': '33',
        'period': FakeDatetime(2025, 2, 6, 11, 0, tzinfo=tzutc()),
        'precip_probability': 20,
        'temperature': -17,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Cloudy',
        'icon_code': '10',
        'period': FakeDatetime(2025, 2, 6, 12, 0, tzinfo=tzutc()),
        'precip_probability': 20,
        'temperature': -17,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Cloudy',
        'icon_code': '10',
        'period': FakeDatetime(2025, 2, 6, 13, 0, tzinfo=tzutc()),
        'precip_probability': 20,
        'temperature': -16,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Cloudy',
        'icon_code': '10',
        'period': FakeDatetime(2025, 2, 6, 14, 0, tzinfo=tzutc()),
        'precip_probability': 20,
        'temperature': -16,
        'wind_direction': 'E',
        'wind_speed': 10,
      }),
      dict({
        'condition': 'Snow',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 15, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -15,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 16, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -14,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 17, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -13,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 18, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -12,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 19, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -11,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow. Risk of freezing drizzle',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 20, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -9,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow. Risk of freezing drizzle',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 21, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -8,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow. Risk of freezing drizzle',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 22, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -7,
        'wind_direction': 'E',
        'wind_speed': 15,
      }),
      dict({
        'condition': 'Snow. Risk of freezing drizzle',
        'icon_code': '17',
        'period': FakeDatetime(2025, 2, 6, 23, 0, tzinfo=tzutc()),
        'precip_probability': 100,
        'temperature': -5,
        'wind_direction': 'SW',
        'wind_speed': 10,
      }),
    ]),
    language='english',
    lat=45.33,
    lon=-75.58,
    max_data_age=2,
    metadata=MetaData(attribution='Data provided by Environment Canada', timestamp=FakeDatetime(2025, 2, 5, 23, 8, 27, tzinfo=tzutc()), station="Ottawa Macdonald-Cartier Int'l Airport", location='Ottawa (Kanata - Orléans)', cache_returned_on_update=0, last_update_error=''),
    site_list=list([
      dict({
        'Codes': 's0000001',
        'English Names': 'Athabasca',
        'Latitude': 54.72,
        'Longitude': -113.28,
        'Province Codes': 'AB',
      }),
      dict({
        'Codes': 's0000002',
        'English Names': 'Clearwater',
        'Latitude': 51.65,
        'Longitude': -120.03,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000003',
        'English Names': 'Valemount',
        'Latitude': 52.83,
        'Longitude': -119.25,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000004',
        'English Names': 'Grand Forks',
        'Latitude': 49.03,
        'Longitude': -118.43,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000005',
        'English Names': 'McBride',
        'Latitude': 53.3,
        'Longitude': -120.17,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000006',
        'English Names': 'Merritt',
        'Latitude': 50.11,
        'Longitude': -120.79,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000007',
        'English Names': 'Masset',
        'Latitude': 54.02,
        'Longitude': -132.1,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000008',
        'English Names': 'Invermere',
        'Latitude': 50.52,
        'Longitude': -116.03,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000009',
        'English Names': 'Dome Creek',
        'Latitude': 53.75,
        'Longitude': -121.03,
        'Province Codes': 'BC',
      }),
      dict({
        'Codes': 's0000010',
        'English Names': 'Little Grand Rapids',
        'Latitude': 52.04,
        'Longitude': -95.46,
        'Province Codes': 'MB',
      }),
      dict({
        'Codes': 's0000430',
        'English Names': 'Ottawa (Kanata - Orléans)',
        'Latitude': 45.33,
        'Longitude': -75.58,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000431',
        'English Names': 'Leamington',
        'Latitude': 42.1,
        'Longitude': -82.56,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000432',
        'English Names': 'Ulukhaktok',
        'Latitude': 70.73,
        'Longitude': -117.75,
        'Province Codes': 'NT',
      }),
      dict({
        'Codes': 's0000433',
        'English Names': 'Ingonish',
        'Latitude': 46.69,
        'Longitude': -60.37,
        'Province Codes': 'NS',
      }),
      dict({
        'Codes': 's0000434',
        'English Names': 'Smiths Falls',
        'Latitude': 44.9,
        'Longitude': -76.02,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000435',
        'English Names': 'Sharbot Lake',
        'Latitude': 44.77,
        'Longitude': -76.69,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000436',
        'English Names': 'Kemptville',
        'Latitude': 45.02,
        'Longitude': -75.64,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000437',
        'English Names': 'Kirkland Lake',
        'Latitude': 48.15,
        'Longitude': -80.03,
        'Province Codes': 'ON',
      }),
      dict({
        'Codes': 's0000438',
        'English Names': 'Windsor',
        'Latitude': 44.99,
        'Longitude': -64.14,
        'Province Codes': 'NS',
      }),
      dict({
        'Codes': 's0000439',
        'English Names': 'Kentville',
        'Latitude': 45.08,
        'Longitude': -64.5,
        'Province Codes': 'NS',
      }),
      dict({
        'Codes': 's0000440',
        'English Names': 'Lunenburg',
        'Latitude': 44.38,
        'Longitude': -64.31,
        'Province Codes': 'NS',
      }),
    ]),
    station_id=tuple(
      'ON',
      '430',
    ),
  )
# ---
